### PR TITLE
Stress test

### PR DESCRIPTION
**Pull request by ashwinNV**
I think I've tested everything except actually using it with CallTestScript.py. Works with Times and Frames, and detects growth from the start frame.

### CODE REVIEW

File: Framework/Source/Framework.h [hunk 1]
```diff
@@ -31,6 +31,7 @@
 #include <memory>
 #include "glm/glm.hpp"
 #include <iostream>
+#include <string>
```
Comment (nbenty) [inline Framework/Source/Framework.h]: Why is here? Framework.h is included by every file in the solution, we should try to minimize the amount of headers we put here.
Move it to whichever file that actually requires it.

File: Framework/Source/Sample.cpp [hunk 2]
```diff
@@ -40,504 +40,695 @@
...
-                initVideoCapture();
-            }
-            else if(!keyEvent.mods.isAltDown && !keyEvent.mods.isCtrlDown && !keyEvent.mods.isShiftDown)
-            {
-                switch(keyEvent.key)
-                {
-                case KeyboardEvent::Key::F12:
-                    mCaptureScreen = true;
-                    break;
+	Sample::Sample()
```
Comment (nbenty) [inline Framework/Source/Sample.cpp]: What happened to all the spaces in this file? Are your tabs settings correct?

File: Framework/Source/Sample.cpp [hunk 3]
```diff
@@ -40,504 +40,695 @@
...
+	//	Capture the Current Memory and write it to the provided memory check.
+	void Sample::getMemoryStatistics(MemoryCheck & memoryCheck)
+	{
+		memoryCheck.pTotalVirtualMemory = getTotalVirtualMemory();
+		memoryCheck.pTotalUsedVirtualMemory = getUsedVirtualMemory();
+		memoryCheck.pCurrentlyUsedVirtualMemory = getProcessUsedVirtualMemory();
+	}
+
+	//	Write the Memory Check Range, in terms of Frames to a file. Outputs Difference, Start and End Frames and Memories.
+	void Sample::writeMemoryFrameRange(const MemoryCheckRange & memoryCheckRange)
```
Comment (nbenty) [inline Framework/Source/Sample.cpp]: This doesn't belong here. This should be in SampleTest

File: Framework/Source/Sample.cpp [hunk 4]
```diff
@@ -40,504 +40,695 @@
...
+		std::string endCUVM_MB = std::to_string(memoryCheckRange.endCheck.pCurrentlyUsedVirtualMemory / (1024 * 1024));
+
+		std::string endCheck = "At the End Frame, " + std::to_string(memoryCheckRange.endCheck.pFrame) + ", : \n ";
+		endCheck = endCheck + ("Total Virtual Memory : " + endTVM_B + " bytes, " + endTVM_MB + " MB. \n");
+		endCheck = endCheck + ("Total Used Virtual Memory By All Processes : " + endTUVM_B + " bytes, " + endTUVM_MB + " MB. \n");
+		endCheck = endCheck + ("Virtual Memory used by this Process : " + endCUVM_B + " bytes, " + endCUVM_MB + " MB. \n \n");
+
+		//	Compute the Difference Between the Two.
+		std::string differenceCheck = "Difference : \n";
+		long long difference = 0;
```
Comment (nbenty) [inline Framework/Source/Sample.cpp]: we don't use `long long`. Use uint64_t instead.

File: Framework/Source/Sample.cpp [hunk 5]
```diff
@@ -40,504 +40,695 @@
...
+		std::string filename = getExecutableName();
+
+		// Now we have a folder and a filename, look for an available filename (we don't overwrite existing files)
+		std::string prefix = std::string(filename) + ".MemoryFrameCheck";
+		std::string executableDir = getExecutableDirectory();
+		std::string txtFile;
+
+
+		//	Get an available filename.
+		if (findAvailableFilename(prefix, executableDir, "txt", txtFile))
```
Comment (nbenty) [inline Framework/Source/Sample.cpp]: This is confusing. We use the same naming convention for log file. Give it a unique name (like prefix with 'mem' or something similar)

File: Framework/Source/Sample.cpp [hunk 6]
```diff
@@ -40,504 +40,695 @@
...
+		else
+		{
+			//	Log Error.
+			logError("Could not find available filename when checking memory.");
+		}
+
+	}
+
+	//	Write the Memory Check Range, in terms of Time to a file. Outputs Difference, Start and End Times and Memories.
+	void Sample::writeMemoryTimeRange(const MemoryCheckRange & memoryCheckRange)
```
Comment (nbenty) [inline Framework/Source/Sample.cpp]: This function looks very similar to the one above. Can't you abstract them for better code sharing?

Comment (nbenty) [inline Framework/Source/Sample.cpp]: This function looks very similar to the one above. Can you abstract them for better code sharing?

File: Framework/Source/Sample.cpp [hunk 7]
```diff
@@ -40,504 +40,695 @@
...
+			of.open(txtFile);
+			of << keystring;
+			of << differenceCheck;
+			of << startCheck;
+			of << endCheck;
+			of.close();
+		}
+		else
+		{
+			//	Log Error.
```
Comment (nbenty) [inline Framework/Source/Sample.cpp]: More space issues

File: Framework/Source/Sample.h [hunk 8]
```diff
@@ -161,17 +161,60 @@ namespace Falcor
         float mCurrentTime = 0;                           ///< Global time
         ArgList mArgList;
 
+		//	The Memory Check for one frame.
+		struct MemoryCheck
+		{
```
Comment (nbenty) [inline Framework/Source/Sample.h]: That doesn't belong here, plus you have space issues.

File: Framework/Source/Sample.h [hunk 9]
```diff
@@ -161,17 +161,60 @@ namespace Falcor
         float mCurrentTime = 0;                           ///< Global time
         ArgList mArgList;
 
+		//	The Memory Check for one frame.
+		struct MemoryCheck
+		{	
+			unsigned long pFrame;
```
Comment (nbenty) [inline Framework/Source/Sample.h]: Please look at the coding convention document for how we use variables. We don't use the regular C types, we use int32_t, uint32_t, etc.

File: Framework/Source/Sample.h [hunk 10]
```diff
@@ -161,17 +161,60 @@ namespace Falcor
...
     protected:
         void renderFrame() override;
         void handleWindowSizeChange() override;
         void handleKeyboardEvent(const KeyboardEvent& keyEvent) override;
         void handleMouseEvent(const MouseEvent& mouseEvent) override;
         virtual float getTimeScale() final { return mTimeScale; }
         void initVideoCapture();
-        void captureScreen();
+	
+		/**	Capture the Current Memory and write it to the provided memory check.
```
Comment (nbenty) [inline Framework/Source/Sample.h]: Why is that part of Sample? Shouldn't it be a utility function is OS.h where the sample test can just call?

File: Framework/Source/SampleTest.h [hunk 11]
```diff
@@ -76,6 +76,7 @@ namespace Falcor
 
         enum class TaskType
         {
+			MemoryCheck,
```
Comment (nbenty) [inline Framework/Source/SampleTest.h]: Space/tabs

File: Test/MemoryTestList.txt [hunk 12]
```diff
@@ -0,0 +1,3 @@
+[../Falcor.sln {Released3d12 ..\Bin\x64\Release\}
+ShaderBuffers {-test -memchecktimes 240-3840 -ssframes 50 -shutdowntime 4000} {released3d12}
```
Comment (nbenty) [inline Test/MemoryTestList.txt]: I don't care about ShaderBuffer memory check. That should be the feature demo with the sun-temple

File: Framework/Source/Sample.cpp [hunk 13]
```diff
@@ -40,504 +40,695 @@
...
+		}
+		else if (!frameTest && endRange && memoryTimeCheckRange.active)
+		{
+			getMemoryStatistics(memoryTimeCheckRange.endCheck);
+			memoryTimeCheckRange.endCheck.pFrame = frameRate().getFrameCount();
+			memoryTimeCheckRange.endCheck.pEffectiveTime = mCurrentTime;
+			writeMemoryTimeRange(memoryTimeCheckRange);
+			memoryTimeCheckRange.active = false;
+		}
+		else
```
Comment (kyaoNV) [inline Framework/Source/Sample.cpp]: Empty else

File: Framework/Source/SampleTest.cpp [hunk 14]
```diff
@@ -30,384 +30,459 @@
...
+		{
+			initFrameTests();
+			initTimeTests();
+			onInitializeTesting();
+		}
+	}
+
+	void SampleTest::beginTestFrame()
+	{	
+		//	Sample Memory Leak :
```
Comment (kyaoNV) [inline Framework/Source/SampleTest.cpp]: Remove test code

File: Test/CallTestScript.py [hunk 15]
```diff
@@ -4,6 +4,7 @@
 from datetime import date
 import shutil
 import stat
+import sys
```
Comment (kyaoNV) [inline Test/CallTestScript.py]: Added an import but nothing else in the file changed?

File: Test/OutputTestingHtml.py [hunk 16]
```diff
@@ -103,29 +113,103 @@ def getImageCompareResultsTable(slnInfo):
...
         for result in slnInfo.systemResultList:
-            if len(result.CompareResults) > 0:
+            if len(result.CompareMemoryFrameResults) > 0:
+                hasFrameCheck = True
+
+        if hasFrameCheck :
+            html += '<table style="width:100%" border="1">\n'
+            html += '<tr>\n'
+            html += '<th colspan=\'' + str(maxCols) + '\'> Memory Frame Checks </th> \n'
+            html += '</tr>\n'
```
Comment (kyaoNV) [inline Test/OutputTestingHtml.py]: Nitpick, but there are a bunch of lines with a lot of trailing whitespace in this file.

**Pull request merged**